### PR TITLE
feat(connection-svc): initial implementation

### PR DESCRIPTION
**Pull request by JeremyTheocharis**
Fixes: https://linear.app/united-manufacturing-hub/issue/ENG-2842/write-the-connection-service-as-a-precursor-for-protocol-converter

This is the initial implementation for the connection service with all unit tests passing.

How I implemented this in 2 hours:
1. During the check-in I wrote down a rough implementation plan
2. That implementation plan I then let verify by chatgpt and detail it out (https://chatgpt.com/share/6800e60b-88f4-8010-8b5f-24fce585eb17). I asked it to ask questions, which I then answered. I also explained the full business logic behind it. Took me 30-45min
3. I gave cursor the below prompt together with the implementation plan from chatgpt
4. I then needed to fix the initial implementation, namely I told it to make it similar to the other services. First I told it to make it similar to the file and folder structure. then i tasked it to write tests simialr to the others (ginkgo v2/ gomega). Then I checked the first test and realzied the AI was cheating as it was testing against a service mock, so I told it to sue teh nmap mock against.
5. I ran the tests which resulted in failures
6. I started teh debugger and went through the flow step by step, copied soemtiems the intermediary variables to the LLm and asked to explain the issue to me
7. Once it was working, and I was confident that I am testign what needs to be tested, I prompted it to create a testing plan, simialr to the otehr services, for the entire business lgoic.
8. then I tasked it to implement the busoness logi
9. repeated steps 5 - 6 for the new tests
10. Now I tasked it to overhaul the documentation of that package, namely to add the business logic (which is all in the tests) to the variables, constructors, etc., so that the next AI when building on top of the connection service, knows how the business logic is working
11. I am now writing this PR and waiting for your review and coderabbit review (probably missed some things here)

## Implementation prompt for cursor
```
I have now this technical implementation plan and I want to execute it step by step from 1.1 to 1.4

I will task you in implementing this, but before you do that, can you fetch the codebase, find all the relevant files, etc. and check whether you have everything to get started with implementing steps 1.1 - 1.4? if something is missing or unclear or even impossible or will not work, ask me as many questions as you like. if you understood it, feel free to write "UNDERSTOOD"
```

## Summary by CodeRabbit

- **New Features**
  - Introduced a new Connection Service for monitoring and managing network connectivity, including health and flakiness detection.
  - Added configuration options for managing multiple connection services at runtime.
  - Added constants and logging components for enhanced connection state management and observability.

- **Bug Fixes**
  - Improved mock service behavior to better simulate per-service states in testing.

- **Tests**
  - Added comprehensive unit tests for the Connection Service, including flakiness detection and edge cases.
  - Introduced mock implementations and Ginkgo test suite setup for robust test coverage.

- **Chores**
  - Added a debug configuration for running Go tests with Ginkgo in the service-connection package.

### CODE REVIEW

File: umh-core/pkg/config/connectionconfig/config.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// ConnectionConfig is referenced from FullConfig.Internal.Connection.
+type ConnectionConfig struct {
+	Name                    string                  `json:"name"`
+	Type                    string                  `json:"type"`            // "nmap" (default) – future: "icmp", …
+	DesiredFSMState         string                  `json:"desiredFSMState"` // "up"|"stopped"
+	ConnectionServiceConfig ConnectionServiceConfig `json:"serviceConfig"`
+}
+
+// ConnectionServiceConfig holds the configuration for the connection service
+type ConnectionServiceConfig struct {
```
Comment (Scarjit) [inline umh-core/pkg/config/connectionconfig/config.go:26]: This struct seems unused

File: umh-core/pkg/config/connectionconfig/config.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	// For Nmap:
+	Hostname string `json:"hostname"`
+	Port     uint16 `json:"port"`
+	Protocol string `json:"protocol"` // "tcp"|"udp"
```
Comment (Scarjit) [inline umh-core/pkg/config/connectionconfig/config.go:30]: This would be a good place for an enum

File: umh-core/pkg/service/connection/connection.go [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+	"go.uber.org/zap"
+)
+
+// ServiceInfo holds information about the connection's health status.
+// It uses separate boolean flags for different health aspects instead of
+// a single status value, making it easier to check specific conditions.
+type ServiceInfo struct {
+	// IsRunning indicates if the underlying Nmap service is running.
+	// This refers to the monitoring service itself, not the target endpoint.
+	IsRunning bool
```
Comment (Scarjit) [inline umh-core/pkg/service/connection/connection.go:44]: IsRunning, PortStateOpen, IsReachable, IsFlaky would make an good enum.
 
enum (
 Down = iota
 Running
 Reachable
 Flaky
)

(PortStateOpen is not needed here, as there shouldnt be any situation where IsRunning == false && PortStateOpen == true).

File: umh-core/pkg/service/connection/connection.go [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+	hasFiltered := false
+
+	for _, scan := range scans {
+		if scan.NmapStatus.LastScan == nil {
+			continue
+		}
+
+		state := scan.NmapStatus.LastScan.PortResult.State
+		switch state {
+		case "open":
```
Comment (Scarjit) [inline umh-core/pkg/service/connection/connection.go:497]: These would be nice enums

File: umh-core/pkg/service/connection/connection_test.go [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+				Expect(serviceInfo.IsReachable).To(BeFalse())
+
+				// Third call - open again
+				mockNmap.SetServicePortState(connName, "open", 25.0)
+				serviceInfo, err = connService.Status(ctx, nil, connName, tick)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(serviceInfo.IsReachable).To(BeTrue())
+			})
+
+			JustBeforeEach(func() {
```
Comment (Scarjit) [inline umh-core/pkg/service/connection/connection_test.go:423]: Is this really needed if all you have is a single It node ?

Comment (Scarjit) [reply umh-core/pkg/service/connection/connection_test.go:423]: There are also multiple other occurences of this.

**Review by Scarjit — CHANGES_REQUESTED** [review]

File: umh-core/pkg/config/connectionconfig/config.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectionconfig
+
+// ConnectionConfig is referenced from FullConfig.Internal.Connection.
+type ConnectionConfig struct {
```
Comment (led0nk) [inline umh-core/pkg/config/connectionconfig/config.go:18]: why do we need that? we already have a ConnectionConfig in config.go (from you pr here) which we should use.
In there if you want for "future: icmp" we can add something to the ServiceInfo"

File: umh-core/pkg/config/connectionconfig/config.go (hunk 1 above)
Comment (led0nk) [reply umh-core/pkg/config/connectionconfig/config.go:26]: this should already exist in package.go from `connectionserviceconfig` and has `NmapServiceConfig` embedded, where we could later on add something

File: umh-core/pkg/config/connectionconfig/config.go (hunk 2 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/config/connectionconfig/config.go:30]: needs to be removed anyway

File: umh-core/pkg/service/connection/connection.go (hunk 3 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection.go:44]: also thought about that, and even added it to the comments. but then decided to follow the AI approach and not use an enum and instead multiple boolean flags.

the resulting checks could then be made much easier with IsPortStateOpen && !IsNotFlaky

this could be a little bit better readable as if I were to do it here in an enum

File: umh-core/pkg/config/connectionconfig/config.go (hunk 1 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/config/connectionconfig/config.go:26]: yes, typical AI error, have removed that

File: umh-core/pkg/config/connectionconfig/config.go (hunk 6 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/config/connectionconfig/config.go:18]: typcial AI mistake, removed

File: umh-core/pkg/service/connection/connection_test.go (hunk 5 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection_test.go:423]: chatgpt insists that this is best-practice, and I dont have a strong enough opinion here to do it differently

```

While technically you could combine everything into a single `It` block when there's only one test case, keeping the Ginkgo structure with separate `BeforeEach`, `JustBeforeEach`, and `AfterEach` blocks is still valuable for:

1. Maintaining consistent test structure across the file
2. Separating setup from assertions, improving readability
3. Making it easier to add more test cases in the future
4. Clearly showing the test's phases (setup → execution → verification → cleanup)

The current structure follows Ginkgo best practices where `BeforeEach` handles setup, `JustBeforeEach` runs the function under test, and `AfterEach` handles cleanup, even if there's only one assertion.

```

File: umh-core/pkg/service/connection/connection.go (hunk 4 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection.go]: i think this is more of a style thing, and we already merged nmap, so let me stay here consistent with nmap

**Review by Scarjit — APPROVED** [review]

File: umh-core/pkg/constants/connection.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+const (
+	// ConnectionExpectedMaxP95ExecutionTimePerInstance means that an instance will not reconcile if not enough time is left
+	// Note: in the integration test, we defined an alerting threshold of 80% of the max ticker time, which is 100ms
+	// So by setting this appropriately, we can ensure that an instance will never start if it triggers the alerting threshold
+	ConnectionExpectedMaxP95ExecutionTimePerInstance = time.Millisecond * 20
+)
+
+const (
+	ConnectionUpdateObservedStateTimeout = time.Millisecond * 2
```
Comment (led0nk) [inline umh-core/pkg/constants/connection.go:27]: why not 5 like the others? if there's a reason, might make sense to add a comment on this

File: umh-core/pkg/service/connection/connection.go [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+	fs filesystem.Service,
+	connName string,
+	tick uint64,
+) (ServiceInfo, error) {
+	if ctx.Err() != nil {
+		return ServiceInfo{}, ctx.Err()
+	}
+
+	// Get status from Nmap service
+	nmapStatus, err := c.nmapService.Status(ctx, fs, connName, tick)
```
Comment (led0nk) [inline umh-core/pkg/service/connection/connection.go:242]: I'm currently also unsure about this approach. What happens if the nmap-fsm gets an permanent failure (not coming from nmap-execution)? Might this result in connection being healthy?  Shouldn't we access the fsm's state here to then control the connection fsm based on this?

File: umh-core/pkg/service/connection/connection.go [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+	return cfg.NmapServiceConfig
+}
+
+// updateRecentScans adds a new scan result to the history for flakiness detection.
+// If len(scans) == maxRecentScans (not >=) we drop the oldest element before appending.
+//
+// This history is used by isConnectionFlaky to detect unstable connections
+// that alternate between different port states.
+func (c *ConnectionService) updateRecentScans(connName string, scan nmap.ServiceInfo) {
+	// TODO(perf): replace slice with constant-size ring-buffer O(1) append
```
Comment (led0nk) [inline umh-core/pkg/service/connection/connection.go:467]: if someone else has to impl this TODO, please give more context

File: umh-core/pkg/service/connection/connection.go [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+
+	// Add the new scan (keeping only the most recent ones)
+	scans := c.recentScans[connName]
+	if len(scans) >= c.maxRecentScans {
+		// Remove oldest scan
+		scans = scans[1:]
+	}
+	scans = append(scans, scanCopy)
+	c.recentScans[connName] = scans
+}
```
[the remark this answers is not in the document]
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection.go:505]: not relevant for now

File: umh-core/pkg/service/connection/connection.go (hunk 8 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection.go:242]: very good catch, indeed we shouldnt use the service but instead the nmap manager

File: umh-core/pkg/constants/connection.go (hunk 7 above)
Comment (JeremyTheocharis) [reply umh-core/pkg/constants/connection.go]: I removed them, shouldnt be part of the service anyway

File: umh-core/pkg/service/connection/connection.go [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+		// Skip the config with matching name
+		if config.Name != connName {
+			newConfigs = append(newConfigs, config)
+		}
+	}
+	c.nmapServiceConfigs = newConfigs
+
+	// Clean up in-memory state
+	delete(c.recentScans, connName)
+	delete(c.prevInfo, connName)
```
[the remark this answers is not in the document]
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection.go:465]: no not needed, it is all single threaded

File: umh-core/pkg/service/connection/connection_test.go [hunk 12]
```diff
@@ -0,0 +1,747 @@
...
+				targetFSMState, maxAttempts, currentState))
+		}
+		err, _ := svc.ReconcileManager(ctx, nil, tick)
+		Expect(err).NotTo(HaveOccurred())
+		tick++
+		attempts++
+
+		// Update current state for next iteration check
+		currentState, err = mgr.GetCurrentFSMState(name)
+		Expect(err).NotTo(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection_test.go:80]: WithTimeout doesnt work, it takes ticks and amount not time to reconcile

File: umh-core/pkg/service/connection/connection_test.go [hunk 13]
```diff
@@ -0,0 +1,747 @@
...
+	BeforeEach(func() {
+		ctx, cancel, mockMgr, mockNmapService, connSvc, tick = newConnTestEnv("history-conn")
+		mockMgr, mockNmapService = nmapfsm.NewNmapManagerWithMockedService("history-mgr")
+
+		// Create service with limited history size
+		connSvc = connection.NewDefaultConnectionService(
+			"history-svc",
+			connection.WithNmapManager(mockMgr),
+			connection.WithMaxRecentScans(maxHistorySize),
+		)
```
[the remark this answers is not in the document]
Comment (JeremyTheocharis) [reply umh-core/pkg/service/connection/connection_test.go:616]: added a comment, it is needed because of WithMaxRecentScans
